alpenhorn/archive.py

```diff
@@ -7,39 +9,83 @@
 
     Attributes
     ----------
     file : foreign key
         Reference to the file of which this is a copy.
     node : foreign key
         The node on which this copy lives (or should live).
-    has_file : string
-        Is the node on the file?
-        - 'Y': yes, the node has the file.
+    has_file : enum
+        Is the file on the node?
+        - 'Y': yes, the node has a copy of the file.
         - 'N': no, the node does not have the file.
-        - 'M': maybe: we've tried to copy/erase it, but haven't yet verified.
+        - 'M': maybe: the file copy needs to be re-checked.
         - 'X': the file is there, but has been verified to be corrupted.
     wants_file : enum
         Does the node want the file?
         - 'Y': yes, keep the file around
         - 'M': maybe, can delete if we need space
-        - 'N': no, attempt to delete
+        - 'N': no, should be deleted
         In all cases we try to keep at least two copies of the file around.
+    ready : bool
+        _Some_ StorageNode I/O classes use this to tell other hosts that
+        files are ready for access.  Other I/O classes do _not_ use this
+        field and assess readiness in some other way, so never check this
+        directly; outside of the I/O-class code itself, use
+        StorageNode.io.ready_path() or StorageNode.remote.pull_ready()
+        to determine whether a remote file is ready for I/O.
     size_b : integer
-        Allocated size of file in bytes (calculated as a multiple of 512-byte
-        blocks).
+        Allocated size of file in bytes (i.e. actual size on the Storage
+        medium.)
+    last_update : timestamp
+        The time at which this record was last updated.  This property
+        is automatically updated on save.  Any value explicity set will be
+        ignored.  (You can still manually set last_update in the
+        database by using the update() method, if necessary).
     """
 
     file = pw.ForeignKeyField(ArchiveFile, backref="copies")
     node = pw.ForeignKeyField(StorageNode, backref="copies")
     has_file = EnumField(["N", "Y", "M", "X"], default="N")
     wants_file = EnumField(["Y", "M", "N"], default="Y")
-    size_b = pw.BigIntegerField()
+    ready = pw.BooleanField(default=False)
+    size_b = pw.BigIntegerField(null=True)
+    last_update = pw.TimestampField(default=datetime.datetime.now)
+
+    # Re-implement save() to always update last_update
+    def save(self, force_insert=False, only=None):
+        """Save an ArchiveFileCopy.
+
+        Automatically updates last_update to the current time.
+        """
+
+        # Only update if dirty
+        if self.is_dirty():
+            self.last_update = datetime.datetime.now()
+            if only is not None and ArchiveFileCopy.last_update not in only:
+                only.append(ArchiveFileCopy.last_update)
+        return super().save(force_insert, only)
+
+    # Re-implement update() to add last_update if not given
```

Comment:
Hmmm.   I agree coleifer usually knows what he's talking about, but it would be really nice to have an explanation of _how_ it's a problem.  Is it because we need to deal with the `__data` parameter?

Maybe the best way to do it is in the schema, then.

Reply:
Or we could just update the field manually.  Let me have a think about how important it is to be correct.  If it's very important we should push it down into the schema so that no matter how you go about updating the table it gets updated.  If it's not that important, it's probably better to be cautious here.

Reply:
Well, `ON UPDATE CURRENT_TIMESTAMP` isn't too difficult, so let's do that.  I thought a row-update timestamp column was a fairly normal thing to have in a relational database. It is kind of disappointing that peewee doesn't really support it.

Doesn't work in SQLite, though.  Need to figure out if that needs some more work.

Reply:
Yeah. Much of my fear is that given that he doesn't think it's a good idea that a future peewee change might break what we're doing here. That has bitten us before.

Reply:
Apparently you can use a trigger into SQLite to make it work. A bit of a weird, and I think you need to do it explicitly in the db.